### PR TITLE
[move stdlib] minimize calls to length in vector functions

### PR DESCRIPTION
**Pull request by sblackshear**
## Description 

Calling length in a loop is less gas and wall time efficient than calling it once and using a simple loop.

## Test plan 

Existing unit tests, added one new one for good measure
---

## Release notes

Check each box that your changes affect. If none of the boxes relate to your changes, release notes aren't required.

For each box you select, include information after the relevant heading that describes the impact of your changes that a user might notice and any actions they must take to implement updates. 

- [X] Protocol: lower gas price for `append` and `do!` 
- [ ] Nodes (Validators and Full nodes): 
- [ ] Indexer: 
- [ ] JSON-RPC: 
- [ ] GraphQL: 
- [ ] CLI: 
- [ ] Rust SDK:
- [ ] REST API:

### CODE REVIEW

**Review by dariorussi — COMMENTED** [review]
I believe that if we want this in we have to protocol version the change, right?
also is this a bit overzealous?
adding others for comments

**Comment by sblackshear** [comment]
> I believe that if we want this in we have to protocol version the change, right? 

Yep, will fix

> also is this a bit overzealous? adding others for comments

It made a difference in the wall clock time for FOMO, and that code was not atypical (`append` is super common). I actually thought we landed this change already after that episode

**Review by damirka — APPROVED** [review]
Looks good to me, thank you, sir!

File: crates/sui-framework/packages/move-stdlib/sources/vector.move [hunk 1]
```diff
@@ -86,7 +86,12 @@ public fun reverse<Element>(v: &mut vector<Element>) {
 /// Pushes all of the elements of the `other` vector into the `lhs` vector.
 public fun append<Element>(lhs: &mut vector<Element>, mut other: vector<Element>) {
     other.reverse();
-    while (other.length() != 0) lhs.push_back(other.pop_back());
+    let mut i = 0;
+    let len = other.length();
+    while (i < len) {
+        lhs.push_back(other.pop_back());
+        i  = i + 1
+    };
```
Comment (tnowacki) [inline crates/sui-framework/packages/move-stdlib/sources/vector.move:94]:
```suggestion
    other.do!(|x| lhs.push_back(x));
```

File: crates/sui-framework/packages/move-stdlib/sources/vector.move [hunk 2]
```diff
@@ -185,7 +190,12 @@ public macro fun destroy<$T>($v: vector<$T>, $f: |$T|) {
 public macro fun do<$T>($v: vector<$T>, $f: |$T|) {
     let mut v = $v;
     v.reverse();
-    while (v.length() != 0) $f(v.pop_back());
+    let mut i = 0;
+    let len = v.length();
+    while (i < len) {
+        $f(v.pop_back());
+        i = i + 1;
+    };
```
Comment (tnowacki) [inline crates/sui-framework/packages/move-stdlib/sources/vector.move:198]:
```suggestion
    v.length().do!(|_| $f(v.pop_back()));
```

**Review by tnowacki — APPROVED** [review]
Can be done in terms of other macros that will compile to the same thing, but otherwise lgtm!

File: crates/sui-framework/packages/move-stdlib/sources/vector.move (hunk 1 above)
Comment (tnowacki) [reply crates/sui-framework/packages/move-stdlib/sources/vector.move:94]: Will have to remove the `other.reverse()`  and `other.destroy_empty()` calls

File: crates/sui-framework/packages/move-stdlib/sources/vector.move (hunk 2 above)
Comment (sblackshear) [reply crates/sui-framework/packages/move-stdlib/sources/vector.move:198]: Infinite recursion :)

File: crates/sui-protocol-config/src/lib.rs [hunk 3]
```diff
@@ -201,6 +201,7 @@ const MAX_PROTOCOL_VERSION: u64 = 70;
 // Version 70: Enable smart ancestor selection in testnet.
 //             Enable probing for accepted rounds in round prober in testnet
 //             Add new gas model version to update charging of native functions.
+//             Framework update to improve gas/wall time efficiency of some Move stdlib vector functions
```
Comment (tnowacki) [inline crates/sui-protocol-config/src/lib.rs:204]: You won't be able to land this under 70 after the branch cut, and we will need to bump the version here. 
I'm trying to cherry pick #19894 for Walrus into 70. So we can either roll this into that and cherry-pick it too, or wait until #19894 lands

**Pull request merged**